presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/dereference/PushDownDereferenceThroughFilter.java

```diff
@@ -0,0 +1,117 @@
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.iterative.rule.dereference.DereferencePushdown.validDereferences;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(a_x := a.x, b := b)
+ *      Filter(a.x.y = 5 AND b.m = 3)
+ *          Source(a, b)
+ *  </pre>
+ * to:
+ * <pre>
+ *  Project(a_x := expr, b := b)
+ *      Filter(expr.y = 5 AND b.m = 3)
+ *          Project(a := a, b := b, expr := a.x)
+ *              Source(a, b)
+ * </pre>
+ *
+ * Pushes down dereference projections in project node assignments and filter node predicate.
+ */
+public class PushDownDereferenceThroughFilter
+        implements Rule<ProjectNode>
+{
+    private static final Capture<FilterNode> CHILD = newCapture();
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PushDownDereferenceThroughFilter(TypeAnalyzer typeAnalyzer)
+    {
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return project()
+                .with(source().matching(filter().capturedAs(CHILD)));
+    }
+
+    @Override
+    public Result apply(ProjectNode node, Captures captures, Rule.Context context)
+    {
+        FilterNode filterNode = captures.get(CHILD);
+
+        // Pushdown superset of dereference expressions from projections and filtering predicate
+        List<Expression> expressions = ImmutableList.<Expression>builder()
+                .addAll(node.getAssignments().getExpressions())
+                .add(filterNode.getPredicate())
+                .build();
+
+        Map<DereferenceExpression, Symbol> pushdownDereferences = validDereferences(expressions, context, typeAnalyzer, true);
+
+        if (pushdownDereferences.isEmpty()) {
+            return Result.empty();
+        }
+
+        PlanNode source = filterNode.getSource();
+
+        ProjectNode projectNode = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                source,
+                Assignments.builder()
+                    .putIdentities(source.getOutputSymbols())
+                    .putAll(HashBiMap.create(pushdownDereferences).inverse())
+                    .build());
+
+        ExpressionNodeInliner dereferenceReplacer = new ExpressionNodeInliner(pushdownDereferences.entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, mapping -> mapping.getValue().toSymbolReference())));
+
+        PlanNode newFilterNode = new FilterNode(
+                context.getIdAllocator().getNextId(),
+                projectNode,
+                ExpressionTreeRewriter.rewriteWith(dereferenceReplacer, filterNode.getPredicate()));
```

Comment:
Why not just `ExpressionNodeInliner.replaceExpression(filterNode.getPredicate(), <mappings>)`?